### ISSUE TITLE
[HOLD for payment 2023-05-22] [$2000] Incorrect parsing on formatted text after URL inside link

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#17772]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open New Expensify on iOS/Android/[desktop](https://staging.new.expensify.com/r/68947221)
2. Start a chat
3. Send this message : `[click google.com *here*](google.com)`
4. Notice the hyperlink line under * here * changes after a few seconds

## Expected Result : 

The word * here * should be bold and hyperlinked

## Actual Result : 

The word * here * is bold, but the hyperlink is removed after a few second

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [X] Android / native
- [X] Android / Chrome
- [X] iOS / native
- [X ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [X] MacOS / Desktop

**Version Number:** 1.3.4-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43996225/233531590-9d1b2cb9-1410-456b-b304-72199d68d9f7.mp4

https://user-images.githubusercontent.com/43996225/233531605-ed52e669-ce0f-4641-a3c7-bcda26be8365.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @kerupuksambel
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1681939793066839

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01501b1626b65fa2ee</li>
        <li>Upwork Job ID: 1650579616656465920</li>
        <li>Last Price Increase: 2023-05-01</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Christinadobrzyn (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by MelvinBot** [comment]
@Christinadobrzyn Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by Christinadobrzyn** [comment]
Sorry, I was testing this on Friday but forgot to leave a note. I am able to replicate this on all platforms. 

might be related to - https://github.com/Expensify/App/issues/17665 but I don't think so.

Adding External label.

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01501b1626b65fa2ee

**Comment by MelvinBot** [comment]
Current assignee @Christinadobrzyn is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @mananjadhav (`External`)

**Comment by eh2077** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Incorrect parsing on formatted text after URL inside link, like link below

```md
[click google.com *here*](google.com)
```

### What is the root cause of that problem?

From frontend, the input link

```md
[click google.com *here*](google.com)
```

is translated into following html

```html
<a href="https://google.com" target="_blank" rel="noreferrer noopener">click <a href="https://google.com" target="_blank" rel="noreferrer noopener">google.com</a> <strong>here</strong></a>
```

through method [getParsedComment](https://github.com/Expensify/App/blob/4947508bedc2e901082d805c84498f5f394ef655/src/libs/ReportUtils.js#L914-L926) which calls method [ExpensiMark.replace](https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/ExpensiMark.js#L331). The above link html has an anchor tag inside the outer anchor tag, which is invalid because [Nested links are illegal](http://www.w3.org/TR/html401/struct/links.html#h-12.2.2). The backend has some logic to fix the invalid html and return following html to frontend

```html
<a href="https://google.com" target="_blank" rel="noreferrer noopener">click </a><a href="https://google.com" target="_blank" rel="noreferrer noopener">google.com</a> <strong>here</strong>
```

Let's dig how the above invalid nested links are created from frontend.

First, we apply [markdown-link-to-html](https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/ExpensiMark.js#L73-L93) rule to the input

```md
[click google.com *here*](google.com)
```

and output following html converting bold syntax of the link text at the same time inside method [modifyTextForUrlLinks](https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/ExpensiMark.js#L412-L415)

```html
<a href="https://google.com" target="_blank" rel="noreferrer noopener">click google.com <strong>here</strong></a>
```

Second, the [autolink-to-html](https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/ExpensiMark.js#L113-L130) rule is unexpectedly applied to the output anchor tag above.

So the root cause is that the [regex](https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/ExpensiMark.js#L118) 

```js
const regex = new RegExp(
    `(?![^<]*>|[^<>]*<\\/)([_*~]*?)${URL_REGEX}\\1(?![^<]*(<\\/pre>|<\\/code>))`,
    'gi',
);
```

of [autolink-to-html](https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/ExpensiMark.js#L113-L130) rule matches link inside the text of anchor tag, like

```html
<a href="https://google.com" target="_blank" rel="noreferrer noopener">click google.com <strong>here</strong></a>
```

We have two negative lookahead groups, the starting group `(?![^<]*>|[^<>]*<\/)` and the ending group `(?![^<]*(<\/pre>|<\/code>))`, that possibly skip to match the url `google.com` inside the anchor tag.

Let's have a look at the starting negative lookahead group.

The starting negative lookahead group `(?![^<]*>|[^<>]*<\/)` of autolink-to-html [regex](https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/ExpensiMark.js#L118) doesn't skip to match string

```
google.com <strong>here</strong></a>
```

Both alternatives, `[^<]*>` and `[^<>]*<\/`, don't match the above string. 

The first alternative `[^<]*>`  doesn't match the string because `google.com` is not wrapped by `<` and `>`. 

The second alternative `[^<>]*<\/` doesn't match the string because string ` google.com <s` is not ended with `</`. 

Below picture includes several test cases of the starting negative lookahead group

<img width="800" alt="image" src="https://user-images.githubusercontent.com/117511920/233817818-0e42d38c-1443-488f-b9d3-359af6696d98.png">

Then let's check the ending negative lookahead group `(?![^<]*(<\/pre>|<\/code>))`. The ending negative lookahead group also doesn't match string

```
 <strong>here</strong></a>
```

So, both the starting and the ending negative lookahead group can't skip to match the url `google.com` inside the anchor tag.

### What changes do you think we should make in order to solve the problem?

To fix this issue, we can improve the ending negative lookahead group `(?![^<]*(<\/pre>|<\/code>))` of autolink-to-html [regex](https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/ExpensiMark.js#L118) to skip `<strong>` , `<em>` and `<del>` tags inside the `<a>` tag. So we need to let it skip the match if the substring(after the captured url) includes `<strong>`, `<em>` and `<del>` tags and ends with `</a>`, `</pre>` or `</code>` tags. The new ending negative lookahead group can be

```
(?!([^<]*<\/?(strong|em|del)>)*[^<]*(<\/pre>|<\/code>|<\/a>))
```

The capturing group `([^<]*<\/?(strong|em|del)>)*` captures `<strong>`, `<em>`, and `<del>` tags. So the new ending negative lookahead group can skip to match string

```
google.com <strong>here</strong></a>
```

Below picture includes several test cases of the new ending negative lookahead group

<img width="800" alt="image" src="https://user-images.githubusercontent.com/117511920/233819835-2021eff4-bf20-4981-902e-264c01dfce03.png">

The new [regex](https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/ExpensiMark.js#L118) of autolink-to-html will be

```js
`(?![^<]*>|[^<>]*<\\/)([_*~]*?)${URL_REGEX}\\1(?!([^<]*<\\/?(strong|em|del)>)*[^<]*(<\\/pre>|<\\/code>|<\\/a>))`,
```

### What alternative solutions did you explore? (Optional)

To fix this issue, we can also improve the starting negative lookahead group `(?![^<]*>|[^<>]*<\/)` of autolink-to-html [regex](https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/ExpensiMark.js#L118) to skip `<strong>` , `<em>` and `<del>` tags inside the `<a>` tag. So we need to improve the second alternative  `[^<>]*<\/` to let it match string includes `<strong>`, `<em>` and `<del>` tags and ends with `</a>`. The new second alternative can be

```
([^<>]*<\/?(strong|em|del)>)*[^<>]*<\/a>
```

The capturing group `([^<>]*<\/?(strong|em|del)>)*` captures `<strong>`, `<em>`, and `<del>` tags. So the new second alternative of the negative lookahead group can skip to match string

```
google.com <strong>here</strong></a>
```

The new [regex](https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/ExpensiMark.js#L118) of autolink-to-html will be

```js
`(?![^<]*>|([^<>]*<\\/?(strong|em|del)>)*[^<>]*<\\/a>)([_*~]*?)${URL_REGEX}\\3(?![^<]*(<\\/pre>|<\\/code>))`
```

While this solution will also need to adjust the index of `match[i]` in method [modifyTextForUrlLinks](https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/ExpensiMark.js#L366) because the starting negative lookahead group adds two new capturing groups.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @tylerkaraszewski (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by mananjadhav** [comment]
@Christinadobrzyn I am going to unassign myself. Can you please reapply the label to assign someone else?

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @s77rt (`External`)

**Comment by MelvinBot** [comment]
Current assignee @tylerkaraszewski is eligible for the External assigner, not assigning anyone new.

**Comment by s77rt** [comment]
:wave:

@eh2077 I see you have a proposal already, please allow me some time I will review asap.

**Comment by s77rt** [comment]
@eh2077 Thanks for the proposal. Your RCA is about right. However the proposed fix is not exactly what we are looking for here. Can we make the fix more generic? I'm thinking of two possible solutions:
1. Extend the negative lookahead to take into account other tags
2. Rearrange how rules are called

___

I would really prefer solution 2, this will make the input that goes to autolink rule look like that:
```
<a href="https://google.com" target="_blank" rel="noreferrer noopener">click google.com *here*</a>
```

And we would not have to modify the autolink regex.

___

**Current behaviour:**
1. Parse link
2. Parse bold, italic, strikethrough
3. Parse autolink
4. Parse bold, italic, strikethrough

step 3 fails because of step 2

**Suggested behaviour:**
1. Parse link
2. Parse autolink
3. Parse bold, italic, strikethrough 


or any alternatives as long as the bold, italic, strikethrough are parsed once.

___

You can view the other side of the RCA as `link` rule "corrupting" the input for `autolink`.

**Comment by eh2077** [comment]
@s77rt Thanks for your thorough review!

> Extend the negative lookahead to take into account other tags

While I think I have take all possible tags into account because when parsing `autolink`, we only possibly have tags, like `<pre>`, `<code>`, `<a>`, `<em>`, `<strong>` and `<del>` according to the applied [rules](https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/ExpensiMark.js#L21-L107).

> Rearrange how rules are called

That's a great point. I'll check if the regex rules of ['italic', 'bold', 'strikethrough'] can all work properly in the parsing order you suggested.

**Comment by eh2077** [comment]
@s77rt I tried to parse bold, italic, strikethrough both before and after the parsing of `link` and `autolink`. Both tries failed several test cases. I think the main reason to parse `link` and `autolink` before bold, italic, strikethrough is that characters like `_`, `*`, and `~`, are all legal in url and they'll mess up parsing `link` and `autolink` if parsed before url.

If not parsing bold, italic, strikethrough when parsing link [here](https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/ExpensiMark.js#L405-L416), then we can parse them in the order you suggested
1. Parse link
2. Parse autolink
3. Parse bold, italic, strikethrough

and get following failing tests
<img width="800" alt="image" src="https://user-images.githubusercontent.com/117511920/235194050-3cb1eaf3-c4c5-4c3d-aff7-85f5004ddc7d.png">

To fix those failed tests, I think we need to improve regex rules of bold, italic, strikethrough. Hmm, it looks a little bit tricky to fix them all.

**Comment by s77rt** [comment]
@eh2077 Yeah the idea is to parse bold, italic, strikethrough after anchor tag is formed and not before. There is also an idea of doing this:
1. Parse link -> Parse autolink inside that link -> Parse bold, italic, strikethrough
2. Parse autolink -> Parse bold, italic, strikethrough

Not sure this is the optimal way of doing it, but it may lead to more ideas

**Comment by s77rt** [comment]
Not overdue. Working on proposals...

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $2000

**Comment by WikusKriek** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The hyperlink line under * here * changes after a few seconds, when inserting the message: `[click google.com *here*](google.com)`

### What is the root cause of that problem?
Nested formatting can be a problem when working with regular expressions because it can interfere with the patterns being matched. In this case, if the autolink regex matches a piece of text that already contains formatting tags (e.g. strong, em, del), the resulting HTML may be incorrect or invalid. This is because the autolink replacement function is adding its own formatting tags around the matched text, which may conflict with the existing tags. 

### What changes do you think we should make in order to solve the problem?

We can change the rule ordering.

Current behaviour:

1. Parse link
2. Parse autolink
3. Parse bold, italic, strikethrough

Suggested behaviour:

1. Parse link
2. Parse bold, italic, strikethrough
3. Parse autolink

Having the rules in this order an then calling `replaceAll('<strong>', '').replaceAll('</strong>', '').replaceAll('<em>', '').replaceAll('</em>', '').replaceAll('<del>', '').replaceAll('</del>', '')` in the autolink replacement function, ensures that the resulting HTML will be valid and display correctly.

### What alternative solutions did you explore? (Optional)

Negative lookbehind assertion (?<!<\/a>) to ensure that the formatting characters are not preceded by a closing <a> tag.

**Comment by s77rt** [comment]
@WikusKriek Thanks for the proposal. Your RCA is a little unclear / incomplete. Can you please elaborate?
> This is because the autolink replacement function is adding its own formatting tags around the matched text

This is not necessary a problem, we still need to parse bold, italic, strikethrough rules within matched the autolink text.

**Comment by ahmedGaber93** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Incorrect parsing on formatted text after URL inside link.
### What is the root cause of that problem?
the 'autolink' process [modifyTextForUrlLinks](https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/ExpensiMark.js#L366) work in loop to match all auto links in the text, but there is no case to **prevent matching nested auto links** or to **prevent replace them**. 
```javascript
while (match !== null) {
    // do replace here
    ...
    // Now we move to the next match that the js regex found in the text
    match = regex.exec(textToCheck);
}

// for example for loop result
// text >>> [click google.com *here*](google.com)
// loop 1 >> <a href="https://google.com">click google.com <strong>here</strong></a>
// loop 2 >> <a href="https://google.com">click <a href="https://google.com">google.com</a> <strong>here</strong></a>
```
### What changes do you think we should make in order to solve the problem?
#### prevent replace nested auto links.
```javascript
// add this line here above if (abort) {} else {}

if (replacedText.lastIndexOf('<a') > replacedText.lastIndexOf('</a')){
   // this auto link match is nested 
    abort = true;
}

if (abort) {} else {}
```

### What alternative solutions did you explore? (Optional)
prevent match nested auto links, But this point may have a high level of complexity.

**Comment by eh2077** [comment]
> @eh2077 Yeah the idea is to parse bold, italic, strikethrough after anchor tag is formed and not before. There is also an idea of doing this:
> 
> 1. Parse link -> Parse autolink inside that link -> Parse bold, italic, strikethrough
> 2. Parse autolink -> Parse bold, italic, strikethrough
> 
> Not sure this is the optimal way of doing it, but it may lead to more ideas

@s77rt I spent a bit more time on researching this. Here is the update

I think we don't need to parse autolink inside a link because nested links are illegal, see this [SO discussion](https://stackoverflow.com/questions/9882916/are-you-allowed-to-nest-a-link-inside-of-a-link). 

So the parsing steps should be

1. Parse link -> Parse bold, italic, strikethrough
2. Parse autolink -> Parse bold, italic, strikethrough

which is same with the current implementation. I did try to get this issue fixed with the idea of rearranging the order of parsing but no luck without making changes to existing regex rules.

**Comment by eh2077** [comment]
> @ahmedGaber93 Thanks for the proposal. Your RCA is correct, but I think the suggested fix can be improved, can we move that check into the regex so the text won't even match? I think we can achieve that using lookbehind but it has poor support on iOS Safari so let's try to avoid that.

@s77rt I think what you're looking for here is my original proposal https://github.com/Expensify/App/issues/17772#issuecomment-1520699876. Is it?

**Comment by eh2077** [comment]
> @eh2077 I didn't mean to parse nested links. I meant to execute the rule autolink inside the matched text of link before applying the bold, italic, strikethrough. If we do that the match will fail and we won't parse nested links, and that's the goal.

@s77rt Thanks for the explanation. 

> 
> We still have two approach I suppose:
> 
> 1. Change autolink regex not to match nested links (links that are already inside an anchor tag)

Yes, I agree with it and I proposed a solution in my [proposal](https://github.com/Expensify/App/issues/17772#issuecomment-1520699876).

> 2. Change order of rules so the autolink will simply not match the nested links (this requires to change when to apply bold, italic, strikethrough rules - they must be applied after autolink)

I also have researched this potential solution and I found that we'll need to change regex rules of bold, italic and strikethrough to get things worked. I updated in the [comment](https://github.com/Expensify/App/issues/17772#issuecomment-1527762316) and you might check it if you occasionally missed that comment.

**Comment by ahmedGaber93** [comment]
@s77rt 
> can we move that check into the regex so the text won't even match? I think we can achieve that using lookbehind but it has poor support on iOS Safari.

I think disable matching nested auto links will have a high level of complexity.
In the below case, we can use similar behavior to simulate lookbehind.
```javascript
[click google.com *here*](google.com)
<a>click google.com <strong>here</strong></a>
```
But we need to cover these cases also, not only the case above.
```javascript
// change *here* position.
[click *here* google.com  *here* ](google.com)
<a>click <strong>here</strong> google.com <strong>here</strong> </a>

// put link in bold * google.com *.
[click * google.com * here ](google.com)
<a>click <strong> google.com </strong> </a>
```

use nested strong en del
```javascript

// nested strong en del
[click *_~here~_* google.com *_~here~_* ](google.com)
<a>click <strong><em><del>here</del></em></strong> google.com <strong><em><del>here</del></em></strong> </a>

// before and after auto link
google.com [click google.com *here*](google.com) google.com 
```
and rich text like this
```javascript
[click ~here~ *_~here~_* *_~google.com~_* click google.com othertext *here* *_here_* othertext *_~here~_* othertext](google.com)
<a>
  click
  <del>here</del>
  <strong><em><del>here</del></em></strong>
  <strong><em><del>google.com</del></em></strong> // nested autolink
  click
  google.com  // nested autolink
  othertext
  <strong>here</strong>
  <strong><em>here</em></strong>
  othertext
  <strong><em><del>here</del></em></strong>
  othertext
</a>
```

**Comment by s77rt** [comment]
@eh2077 

> I think what you're looking for here is my original proposal https://github.com/Expensify/App/issues/17772#issuecomment-1520699876. Is it?

Not exactly. Your proposal is based on other tags (`del`, `em`, etc.). I'm looking for a regex that is based on the anchor tag only i.e. do not match urls inside anchor tags.

**Comment by s77rt** [comment]
@ahmedGaber93 Thanks for the follow up. I'm not totally opposed on achieving that using string operations, similarly on how we are doing it on [excluding email addresses](https://github.com/Expensify/expensify-common/blob/3cdaa947fe77016206c15e523017cd50678f2359/lib/ExpensiMark.js#L373C21-L376). However your solution will abort all following urls if any nested url is matched (and aborted), example: `[click google.com *here*](google.com) github.com`.

**Comment by eh2077** [comment]
> @eh2077
> 
> > I think what you're looking for here is my original proposal [#17772 (comment)](https://github.com/Expensify/App/issues/17772#issuecomment-1520699876). Is it?
> 
> Not exactly. Your proposal is based on other tags (`del`, `em`, etc.). I'm looking for a regex that is based on the anchor tag only i.e. do not match urls inside anchor tags.

Hi @s77rt , in the solution part, I said

> To fix this issue, we can improve the ending negative lookahead group `(?![^<]*(<\/pre>|<\/code>))` of autolink-to-html [regex](https://github.com/Expensify/expensify-common/blob/e93e1eb448ad6bdbde911fd6239f70d5e749635e/lib/ExpensiMark.js#L118) to skip `<strong>` , `<em>` and `<del>` tags inside the `<a>` tag. So we need to let it skip the match if the substring(after the captured url) includes `<strong>`, `<em>` and `<del>` tags and ends with `</a>`, `</pre>` or `</code>` tags. The new ending negative lookahead group can be `(?!([^<]*<\/?(strong|em|del)>)*[^<]*(<\/pre>|<\/code>|<\/a>))`

I think my solution does add a negative lookahead group based on the anchor tag(parsed by previous link rule) only. The new ending negative lookahead group I proposed is to skip matching url inside anchor tags. Note that based on the current parsing order, the **anchor tag** matched in the parsing autolink is possible to contain `<strong>` , `<em>` and `<del>` tags inside it. There's possible to have a more optimistic regex but I think I did come up the kind of solution in my proposal.

I'm sorry to restate my proposal again but I'm just worried about we're missing some context of my original proposal because we discussed much about the other possible solution and you maybe multi-threading on many issues😄.

**Comment by ahmedGaber93** [comment]
@s77rt 
>I'm not totally opposed on achieving that using string operations, similarly on how we are doing it on [excluding email addresses](https://github.com/Expensify/expensify-common/blob/3cdaa947fe77016206c15e523017cd50678f2359/lib/ExpensiMark.js#L373C21-L376). However your solution will abort all following urls if any nested url is matched (and aborted), example: [click google.com *here*](google.com) github.com.

My fix using `isNested` will not abort any of the following urls. But using the way of [excluding email addresses](https://github.com/Expensify/expensify-common/blob/3cdaa947fe77016206c15e523017cd50678f2359/lib/ExpensiMark.js#L373C21-L376) will do, this because we use `abort` inside loop, and when is set to true inside loop in the first time, it will still true for all next times in loop, see code below.
```javascript
let abort = false; // variable scope out of while
let i = 0;
while (i < 5) {
    // we need to abort fist time only
    // but because abort default value out of while scope if it set to true in first time, it still true for all next times inside loop
    if (i === 0){
        abort = true;
    }
    
    if(abort) console.log(i);
    i ++;
}

//log result 0 1 2 3 4
```
regarding the code above, [excluding email addresses](https://github.com/Expensify/expensify-common/blob/3cdaa947fe77016206c15e523017cd50678f2359/lib/ExpensiMark.js#L373C21-L376) behavior may have an issue, if we type ⁦`@google.com google.com google.com` the first part with @ will set abort = true; and all the next auto links will abort.
The fix should be.
```javascript
while () {
    abort = false; // <<<<<<<<<<<< default value should be set inside the loop to reset every time loop
    if (some case){
        abort = true;
    }
}
```
But this out of the issue scope, and may need more investigation and more tests, so I will report it in slack, and for this issue we can use different case like my proposal `isNested` instead of `abort`.
```javascript

while () {
    // isNested will define again every time inside the loop
    const isNested = replacedText.lastIndexOf('<a') > replacedText.lastIndexOf('</a>');
    if (abort || isNested) {
        abort = true;
    }
}
```

https://user-images.githubusercontent.com/41129870/235879919-cb6e63ad-7c90-42f5-8eaa-093993246adc.mov

**Comment by s77rt** [comment]
@eh2077 Your regex does the following: Do not match urls if they are next to `strong`, `em` or `del` tags. This is really not self-explanatory and may feel like a workaround.

What I'm looking for (as a regex solution) is: Do not match urls if they are inside `a` tags. This is self-explanatory that we are not looking to match urls inside other urls (prevent nested parsing)

**Comment by eh2077** [comment]
> @eh2077 Your regex does the following: Do not match urls if they are next to `strong`, `em` or `del` tags. This is really not self-explanatory and may feel like a workaround.
> 
> What I'm looking for (as a regex solution) is: Do not match urls if they are inside `a` tags. This is self-explanatory that we are not looking to match urls inside other urls (prevent nested parsing)

@s77rt Thanks for the comment. But sorry, I don't think you understand the new ending negative lookahead regex I proposed very well. The new negative lookahead regex `(?!([^<]*<\/?(strong|em|del)>)*[^<]*(<\/pre>|<\/code>|<\/a>))` added the alternative `<\/a>` in the last group to skip the match if it's inside the an anchor tag. And I explained about why need to handle/skip `strong`, `em` or `del` tags in the last comment

> Note that based on the current parsing order, the anchor tag matched in the parsing autolink is possible to contain `<strong>` , `<em>` and `<del>` tags inside it.

You can check my original proposal for the detail explanation of the new ending negative lookahead regex.

**Comment by s77rt** [comment]
@ahmedGaber93 Right, my bad. That would work, but let's try to fix that at a higher level by not matching the text in the first place.

**Comment by s77rt** [comment]
@eh2077 I understood that. I just feel the dependency of specifying the tags in the regex is not a very good thing. For instance that regex won't work correctly with inline code tag since it's missing in the regex, and we may encounter similar bugs for new tags.

**Comment by s77rt** [comment]
BTW the regex fails for many tags and not just the anchor tag, e.g. `<pre>google.com <strong>here</strong></pre>` this will wrongly match google.com for autolink but this case will never occur since the input will never have parsed bold, italic, strikethrough rules. That's why I'm pushing for the re-arrangement solution.

**Comment by eh2077** [comment]
> BTW the regex fails for many tags and not just the anchor tag, e.g. `<pre>google.com <strong>here</strong></pre>` this will wrongly match google.com for autolink but this case will never occur since the input will never have parsed bold, italic, strikethrough rules. That's why I'm pushing for the re-arrangement solution.

No, it won't because `<strong>here</strong>` inside `<pre>` tag is escaped before processing.

**Comment by s77rt** [comment]
Not because  it's escaped but because it never arrives at this state, it will arrive as an input as `'''google *here*'''` and not `'''google <strong>here</strong>'''`

**Comment by eh2077** [comment]
> Not because it's escaped but because it never arrives at this state, it will arrive as an input as `'''google *here*'''` and not `'''google <strong>here</strong>'''`

Sorry I don't think we support the use case you mentioned above. AFAIK, we don't parse strong, del, strikethrough inside the codefence(aka pre tag in html).

**Comment by s77rt** [comment]
Oops a bad example :sweat_smile:

**Comment by s77rt** [comment]
@eh2077 I have pointed a failing case already
```
[click test.com `here`](google.com) github.com
```

This can be fixed easily but shows the downside of the tags dependency. Can we improve the regex not to relay on such tags?

**Comment by ahmedGaber93** [comment]
@s77rt 
>but let's try to fix that at a higher level by not matching the text in the first place.

yes, I already try it before writing my proposal, but I can't find any regexp can support all cases below like `abort` or `isNested` 

```
[click google.com *here*](google.com) 
[click google.com *here*](google.com) google.com
google.com [click google.com *here*](google.com)
[click google.com *here*](google.com) [click google.com *here*](google.com) 
[click google.com *here*](google.com) google.com [click google.com *here*](google.com) 
[click *_~here~_* google.com *_~here~_* ](google.com)
[click ~here~ *_~here~_* *_~google.com~_* click google.com othertext *here* *_here_* othertext *_~here~_* othertext](google.com)
```

**Comment by s77rt** [comment]
@ahmedGaber93 Thanks! I think @eh2077 proposed regex solution can handle those cases as well. Both solutions are acceptable. I think I'm more inclined towards the regex solution since it prevent us from matching the text in the first place rather than matching it then ignoring.

**Comment by eh2077** [comment]
> What I'm looking for (as a regex solution) is: Do not match urls if they are inside `a` tags. This is self-explanatory that we are not looking to match urls inside other urls (prevent nested parsing)

@s77rt I just figured out a new negative lookahead regex that follows this idea. The new regex is `(?!((?:(?!<a).)+)?<\/a>|[^<]*(<\/pre>|<\/code>))`. There're two alternative groups of the outer negative lookahead group, `((?:(?!<a).)+)?<\/a>` and `[^<]*(<\/pre>|<\/code>)`. 

The first alternative group `((?:(?!<a).)+)?<\/a>` is what you expected I think. See below picture about the test cases and regex explanation

<img width="1158" alt="image" src="https://user-images.githubusercontent.com/117511920/236068514-1b277220-2ec5-4cfe-bf11-6c19678b5afd.png">

**Comment by s77rt** [comment]
@eh2077 Cool! That's what I'm looking for.

So the idea is basically to not match urls if they are followed by `</a>` (because we assume those are nested) with the exception if there is another open `<a` tag (to be sure the matched close tag include the url). Would that be about right?

Can you please compile more tests for that? (and send regex101 url)

**Comment by eh2077** [comment]
@s77rt Great to hear that!

> So the idea is basically to not match urls if they are followed by </a> (because we assume those are nested) with the exception if there is another open <a tag (to be sure the matched close tag include the url). Would that be about right?

Yes, the idea is to check text after the matched url
1. If we find substring `<a` before substring `</a>`, then we assume that the matched url is not nested
2. If we find substring `</a>` before substring `<a`, then we assume that the matched url is nested and skip this match


The [regex101.com](https://regex101.com) does't support sharing by url but you can use below regex 
```js
/test\.com(?!((?:(?!<a).)+)?<\/a>|[^<]*(<\/pre>|<\/code>))/gi
```
and text to test
```
click test.com</a>
click test.com</a

<a href="https://google.com" target="_blank" rel="noreferrer noopener">click test.com <code>here</code></a>
<a href="https://google.com" target="_blank" rel="noreferrer noopener">click test.com <code>here</code></a

click test.com <code>here</code></a>
click test.com <code>here</code></a

click test.com <em>here</em></a>
click test.com <em>here</em></a

click test.com <a href="https://google.com" target="_blank" rel="noreferrer noopener">click test.com <code>here</code></a>
click test.com <a href="https://google.com" target="_blank" rel="noreferrer noopener">click test.com <code>here</code></a

```

**Comment by s77rt** [comment]
@eh2077 Thank you! I would love to get the regex a little bit simplified (in the 2nd negative lookahead) but this is not a blocker.

:ribbon: :eyes: :ribbon: C+ reviewed

cc @tylerkaraszewski

**Comment by tylerkaraszewski** [comment]
I agree that this is not the most straightforward regex but I think it's the best option we have so I say we should accept @eh2077's proposal.

**Comment by s77rt** [comment]
We have a proposal. @tylerkaraszewski Can you please assign @eh2077

**Comment by eh2077** [comment]
Ah, it looks like I prepared the PR a little bit earlier😄. The PR https://github.com/Expensify/expensify-common/pull/529 for expensify-common is ready for review. Please help to review it at your convenience, thanks!

cc @s77rt @tylerkaraszewski

**Comment by s77rt** [comment]
Not overdue. We have a PR for expensify-common https://github.com/Expensify/expensify-common/pull/529

**Comment by eh2077** [comment]
@s77rt @tylerkaraszewski The PR https://github.com/Expensify/App/pull/18697 to update expensify-common is ready for review. Please help to review it at your convenience, thanks!

**Comment by tylerkaraszewski** [comment]
It's merged, Melvin.

**Comment by s77rt** [comment]
- The PR that introduced the bug has been identified: This seems to be a bug since day one. I found this [PR](https://github.com/Expensify/expensify-common/pull/264) that seems to party (unintentionally?) fix the nested links issue.
- The offending PR has been commented on: n/a
- A discussion in #expensify-bugs has been started: https://expensify.slack.com/archives/C049HHMV9SM/p1684151092856929
- Determine if we should create a regression test for this bug: No, we have automated tests for this.
